### PR TITLE
openssl_privatekey: add support for format option

### LINKED ISSUE #59253: Allow to select openssl_privatekey format
##### SUMMARY
Allow to specify a new "format" parameter in "openssl_privatekey". Today the output format is decided by a very simple heuristic which requires further commands to work properly.
I need, for example, to generate a RSA private key but in PKCS8 format. The current heuristic uses OpenSSH format for RSA keys and only uses PKCS8 for Ed25519 and similar. This makes me use a "command" with "openssl" to convert the generated key, which leads to idempotency problems and unnecessary complexity.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
openssl_privatekey

##### ADDITIONAL INFORMATION

The current heuristic should be kept unless overwritten by user.

```yaml
openssl_privatekey:
  path: /etc/ssl/private/ansible.com.pem
  format: pkcs8
```
The values allowed are `pkcs1`, `raw` and `pkcs8`

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Fixes #59253.

~~WIP since some things are missing:~~
 * ~~idempotency~~
 * ~~integration tests~~

##### ISSUE TYPE
- New Feature Pull Request

##### COMPONENT NAME
openssl_privatekey

### CODE REVIEW

**Comment by felixfontein** [comment]
What's missing is a proper detection what the current format of a PEM key is (https://github.com/ansible/ansible/pull/60388/files#diff-f47facfb3118315473f7ea281891f42fR649). Everything else is there.

**Comment by felixfontein** [comment]
(During rebase, I squashed to commits so far.)

I probably won't have time to work on this before 2.9 feature freeze, so if anyone is interested in getting this into Ansible 2.9, please help with the key format identification part :)

**Comment by felixfontein** [comment]
I now added a proper key format identification (based on the OpenSSL code, thanks @bkmeneguello for the pointer), and some more tests.

**Comment by bkmeneguello** [comment]
Amazing work, I'll review ASAP

**Comment by felixfontein** [comment]
I'm right now adding another option `format_mismatch` (default value `regenerate`, also offers `convert`) which allows to convert keys to other formats (if possible) instead of regenerating them.

**Comment by felixfontein** [comment]
If sanity checks don't have any surprises, I should be done for now. At least locally, the tests ran through :)

**Comment by felixfontein** [comment]
Tests pass. \o/

ready_for_review

File: lib/ansible/modules/crypto/openssl_privatekey.py [hunk 1]
```diff
@@ -123,6 +123,31 @@
         default: auto
         choices: [ auto, cryptography, pyopenssl ]
         version_added: "2.8"
+    format:
+        description:
+            - Determines which format the private key is written in. By default, PKCS1 (traditional OpenSSL format)
+              is used for all keys which support it. Please note that not every key can be exported in any format.
+            - The value C(auto) selects a backend depending on the key format. The value C(auto_ignore) does the same,
```
Comment (Shaps) [inline lib/ansible/modules/crypto/openssl_privatekey.py]:
```suggestion
            - The value C(auto) selects a format based on the key. The value C(auto_ignore) does the same,
```

File: lib/ansible/modules/crypto/openssl_privatekey.py [hunk 2]
```diff
@@ -123,6 +123,31 @@
...
+              but for existing private key files, it will not force a regenerate when its format is not the automatically
+              selected one for generation.
+            - Note that if the format for an existing private key mismatches, the key is *regenerated* by default.
+              To change this behavior, use the I(format_mismatch) option.
+            - Only supported by the C(cryptography) backend.
```
Comment (Shaps) [inline lib/ansible/modules/crypto/openssl_privatekey.py]: I would be more specific on what is only supported by the cryptography backend

File: lib/ansible/modules/crypto/openssl_privatekey.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/crypto/openssl_privatekey.py]: Good catch! Thanks :)

File: lib/ansible/modules/crypto/openssl_privatekey.py (hunk 2 above)
Comment (felixfontein) [reply lib/ansible/modules/crypto/openssl_privatekey.py]: Good idea. I've extended the text, and also added what the `pyopenssl` backend does (fail if value is not `auto_ignore`).

**Review by bkmeneguello — APPROVED** [review]

**Comment by Shaps** [comment]
shipit

**Comment by felixfontein** [comment]
@bkmeneguello @Shaps thanks a lot for reviewing this! :)

**Pull request merged**